### PR TITLE
feature: add exec timeout support

### LINKED ISSUE #2916: [feature request] exec with timeout
## Why you need it?

exec a process with timeout. after timeout, kill this process.

## How it could be?

expose the OCI kill interface. If timeout, kill the process inside the container.

### PR DESCRIPTION
**Pull request by ZYecho**
Signed-off-by: zhangyue <zy675793960@yeah.net>

### Ⅰ. Describe what this PR did
if exec hang， we need a way to to clean up the resource.
use ctx to judge if timeout， if happen，try to kill with SIGKILL first， then try to kill with SIGTERM

### Ⅱ. Does this pull request fix one issue?

fix #2916 

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
wait for CI

### Ⅳ. Describe how to verify it
wait for CI

### Ⅴ. Special notes for reviews
None

### CODE REVIEW

File: ctrd/container.go [hunk 1]
```diff
@@ -140,47 +139,74 @@ func (c *Client) execContainer(ctx context.Context, process *Process) error {
...
+					logrus.Errorf("failed to kill the exec process, %v", err)
+				}
 			}
-		}
-
-		// success to start which means the cmd is valid and wait
-		// for process.
-		if msg == nil {
+			// Wait for process to be killed
 			status := <-exitStatus
```
Comment (ZYecho) [inline ctrd/container.go]: if failed to kill the execProcess, is it possible to block at this line?

File: ctrd/container.go [hunk 2]
```diff
@@ -137,50 +132,68 @@ func (c *Client) execContainer(ctx context.Context, process *Process) error {
...
+			if err := execProcess.Kill(ctx, syscall.SIGTERM); err != nil && !errdefs.IsNotFound(err) {
+				logrus.Errorf("failed to kill the exec process, %v", err)
 			}
 		}
-
-		// delete the finished exec process in containerd
-		if _, err := execProcess.Delete(context.TODO()); err != nil {
-			logrus.Warnf("failed to delete exec process %s: %s", process.ExecID, err)
+		// Wait for process to be killed
+		status := <-exitStatus
```
Comment (ZYecho) [inline ctrd/container.go]: if failed to kill the execProcess, is it possible to block at this line?

File: ctrd/container.go [hunk 3]
```diff
@@ -137,50 +132,68 @@ func (c *Client) execContainer(ctx context.Context, process *Process) error {
...
 	}
 
-	errCh := make(chan error, 1)
-	defer close(errCh)
-
-	go func() {
-		var msg *Message
+	var msg *Message
+	defer func() {
+		if msg != nil {
```
Comment (zhuangqh) [inline ctrd/container.go]: msg won’t be nil here. or use fast return `if msg == nil return` to reduce nested level.

File: ctrd/container.go [hunk 4]
```diff
@@ -137,50 +132,68 @@ func (c *Client) execContainer(ctx context.Context, process *Process) error {
...
+			//try to force kill the exec process
+			if err := execProcess.Kill(ctx, syscall.SIGTERM); err != nil && !errdefs.IsNotFound(err) {
+				logrus.Errorf("failed to kill the exec process, %v", err)
 			}
 		}
-
-		// delete the finished exec process in containerd
-		if _, err := execProcess.Delete(context.TODO()); err != nil {
-			logrus.Warnf("failed to delete exec process %s: %s", process.ExecID, err)
+		// Wait for process to be killed
```
Comment (zhuangqh) [inline ctrd/container.go]: comment started with lowercase character.

File: ctrd/container.go [hunk 5]
```diff
@@ -137,50 +132,68 @@ func (c *Client) execContainer(ctx context.Context, process *Process) error {
...
+			err:      status.Error(),
+			exitCode: status.ExitCode(),
+			exitTime: status.ExitTime(),
+		}
+	case <-timeCh:
+		// Ignore the not found error because the process may exit itself before kill
+		if err := execProcess.Kill(ctx, syscall.SIGKILL); err != nil && !errdefs.IsNotFound(err) {
+			//try to force kill the exec process
+			if err := execProcess.Kill(ctx, syscall.SIGTERM); err != nil && !errdefs.IsNotFound(err) {
+				logrus.Errorf("failed to kill the exec process, %v", err)
```
Comment (zhuangqh) [inline ctrd/container.go]: we should return error here instead of warning?  cc @fuweid

Comment (fuweid) [reply ctrd/container.go]: +1

**Comment by ZYecho** [comment]
@fuweid @zhuangqh PTAL

**Pull request merged**